[DX-1190] Release Notes 5.3 Tyk Gateway (#4038)
TykTechnologies/tyk-docs: tyk-docs/content/product-stack/tyk-gateway/release-notes/version-5.3.md

---
title: Tyk Gateway 5.3 Release Notes
date: 2023-09-27T15:49:11Z
description: "Release notes documenting updates, enhancements, and changes for Tyk Gateway versions within the 5.3.X series."
tags: ["Tyk Gateway", "Release notes", "v5.3", "5.3.0", "5.3", "changelog"]
---

<!-- Required. oss or licensed. Choose one of the following:
    **Licensed Protected Product**
    Or
    ****Open Source** ([Mozilla Public License](https://github.com/TykTechnologies/tyk/blob/master/LICENSE.md))**
-->

**Open Source** ([Mozilla Public License](https://github.com/TykTechnologies/tyk/blob/master/LICENSE.md))

**This page contains all release notes for version 5.3.X displayed in a reverse chronological order**

## Support Lifetime
<!-- Required. replace X.Y with this release and set the correct quarter of the year -->
Our minor releases are supported until our next minor comes out.

---

## 5.3.0 Release Notes

### Release Date DD Mon YYYY <<update>>

### Breaking Changes
<!-- Required. Use the following statement if there are no breaking changes, or explain if there are -->
##### Attention: Please read this section carefully

> As Tyk OAS transitions out of [Early Access]({{< ref "frequently-asked-questions/using-early-access-features" >}}), we cannot guarantee backward compatibility for versions pre-5.3.0. When upgrading to 5.3.0, Tyk will automatically migrate existing Tyk OAS APIs to the [latest API definition format]({{< ref "tyk-apis/tyk-gateway-api/oas/x-tyk-oas-doc" >}}).

Reviewer: ```suggestion
##### Tyk OAS APIs not backward compatible with pre-5.3.0 Gateways

As Tyk OAS transitions out of [Early Access]({{< ref "frequently-asked-questions/using-early-access-features" >}}), we cannot guarantee backward compatibility for versions pre-5.3.0. When upgrading to 5.3.0, Tyk will automatically migrate existing Tyk OAS APIs to the [latest API definition format]({{< ref "tyk-apis/tyk-gateway-api/oas/x-tyk-oas-doc" >}}).
```

Tyk OAS APIs using the 5.3.0 API definition may not work with pre-5.3.0 versions of Tyk Gateway and doing so could provoke unpredictable behaviour including potential crashes. We recommend retaining copies of existing Tyk OAS API definitions before upgrading in case you choose to downgrade later.

Users are strongly advised to follow the recommended [upgrade instructions](#upgrading-tyk) provided by Tyk before applying any updates.

> 

Reviewer: ```suggestion

##### Python plugin support
Starting from Tyk Gateway version v5.3.0, Python is no longer bundled with the official Tyk Gateway Docker image to reduce exposure to security vulnerabilities in the Python libraries.

Whilst the Gateway still supports Python plugins, you must [extend the image]({{< ref "plugins/supported-languages/rich-plugins/python/python#install-the-python-development-packages" >}}) to add the language support.
```

In this release, we fixed automated token trimming in Redis, ensuring efficient management of OAuth tokens by implementing a new hourly job within the Gateway and providing a manual trigger endpoint. 
</details>
</li>
<li>
<details>
<summary>Tyk Gateway now validates RFC3339 Date-Time Formats</summary>

We fixed a bug in the Tyk OAS Validate Request middleware where we were not correctly validating date-time format schema, which could lead to invalid date-time values reaching the upstream services.
</details>
</li>
<li>
<details>
<summary>Inaccurate Distributed Rate Limiting (DRL) behavior on Gateway startup</summary>

Fixed an issue when using the Distributed Rate Limiter (DRL) where the Gateway did not apply any rate limit until a DRL notification was received. Now the rate of requests will be limited at 100% of the configured rate limit until the DRL notification is received, after which the limit will be reduced to an even share of the total (i.e. 100% divided by the number of Gateways) per the rate limit algorithm design.
</details>
</li>
<li>
<details>
<summary>Duplicate fields added by OAS-to-UDG translator</summary>

> Fixed an issue where the OAS-to-UDG converter was sometimes adding the 5.same field to an object type many times. This caused issues with the resulting GQL schema and made it non-compliant with GQL specification.

Author: @andyo-tyk @agata-wit Is 5.same a typo?

Reviewer: Yes

---
title: Tyk Gateway 5.3 Release Notes
date: 2023-09-27T15:49:11Z
description: "Release notes documenting updates, enhancements, and changes for Tyk Gateway versions within the 5.3.X series."
tags: ["Tyk Gateway", "Release notes", "v5.3", "5.3.0", "5.3", "changelog"]
---

<!-- Required. oss or licensed. Choose one of the following:
    **Licensed Protected Product**
    Or
    ****Open Source** ([Mozilla Public License](https://github.com/TykTechnologies/tyk/blob/master/LICENSE.md))**
-->

**Open Source** ([Mozilla Public License](https://github.com/TykTechnologies/tyk/blob/master/LICENSE.md))

**This page contains all release notes for version 5.3.X displayed in a reverse chronological order**

## Support Lifetime
<!-- Required. replace X.Y with this release and set the correct quarter of the year -->
Our minor releases are supported until our next minor comes out.

---

## 5.3.0 Release Notes

### Release Date DD Mon YYYY <<update>>

### Breaking Changes
<!-- Required. Use the following statement if there are no breaking changes, or explain if there are -->
**Attention: Please read this section carefully.**

> 

Reviewer 2: @dcs3spp @andyo-tyk I think we need a title here for 2 reasons:
1. in case we have more than one topic (so for other cases, not this release)
2. to guide the reader to keep reading in case they are scanning the page and make sure the topic stands out

Also - we should add a link to the actual upgrade instructions `/product-stack/tyk-gateway/release-notes/version-5.3/#upgrade-instructions`

Author: @andyo-tyk @letzya I've updated and added link to upgrade instructions

In this release, we fixed automated token trimming in Redis, ensuring efficient management of OAuth tokens by implementing a new hourly job within the Gateway and providing a manual trigger endpoint. 
</details>
</li>
<li>
<details>
<summary>Tyk Gateway now validates RFC3339 Date-Time Formats</summary>

We fixed a bug in the Tyk OAS Validate Request middleware where we were not correctly validating date-time format schema, which could lead to invalid date-time values reaching the upstream services.
</details>
</li>
<li>
<details>
<summary>Inaccurate Distributed Rate Limiting (DRL) behavior on Gateway startup</summary>

Fixed an issue when using the Distributed Rate Limiter (DRL) where the Gateway did not apply any rate limit until a DRL notification was received. Now the rate of requests will be limited at 100% of the configured rate limit until the DRL notification is received, after which the limit will be reduced to an even share of the total (i.e. 100% divided by the number of Gateways) per the rate limit algorithm design.
</details>
</li>
<li>
<details>
<summary>Duplicate fields added by OAS-to-UDG translator</summary>

> Fixed an issue where the OAS-to-UDG converter was sometimes adding the 5.same field to an object type many times. This caused issues with the resulting GQL schema and made it non-compliant with GQL specification.

Reviewer: ```suggestion
Fixed an issue where the OAS-to-UDG converter was sometimes adding the same field to an object type many times. This caused issues with the resulting GQL schema and made it non-compliant with GQL specification.
```

---
title: Tyk Gateway 5.3 Release Notes
date: 2023-09-27T15:49:11Z
description: "Release notes documenting updates, enhancements, and changes for Tyk Gateway versions within the 5.3.X series."
tags: ["Tyk Gateway", "Release notes", "v5.3", "5.3.0", "5.3", "changelog"]
---

<!-- Required. oss or licensed. Choose one of the following:
    **Licensed Protected Product**
    Or
    ****Open Source** ([Mozilla Public License](https://github.com/TykTechnologies/tyk/blob/master/LICENSE.md))**
-->

**Open Source** ([Mozilla Public License](https://github.com/TykTechnologies/tyk/blob/master/LICENSE.md))

**This page contains all release notes for version 5.3.X displayed in a reverse chronological order**

## Support Lifetime
<!-- Required. replace X.Y with this release and set the correct quarter of the year -->
Our minor releases are supported until our next minor comes out.

---

## 5.3.0 Release Notes

### Release Date DD Mon YYYY <<update>>

### Breaking Changes
<!-- Required. Use the following statement if there are no breaking changes, or explain if there are -->
**Attention: Please read this section carefully**

#### Tyk OAS APIs not backward compatible with pre-5.3.0 Gateways

> As [Tyk OAS APIs]({{< ref "getting-started/key-concepts/high-level-concepts">}}) transitions out of [Early Access]({{< ref "frequently-asked-questions/using-early-access-features" >}}), we cannot guarantee backward compatibility for versions pre-5.3.0. When upgrading to 5.3.0, Tyk will automatically migrate existing Tyk OAS APIs to the [latest API definition format]({{< ref "tyk-apis/tyk-gateway-api/oas/x-tyk-oas-doc" >}}).

Reviewer 2: ```suggestion
This upgrade transitions Tyk OAS APIs out of [Early Access]({{< ref "frequently-asked-questions/using-early-access-features" >}}) to a [new internal format]({{< ref "tyk-apis/tyk-gateway-api/oas/x-tyk-oas-doc" >}}) as part of their move to General Availability. 

**Please Note:**
- Automatic Conversion: Upgrading automatically updates your Tyk OAS API definitions to the latest format used in v5.3.
- Non-Reversible Change: The format change is one-way within the product. Reverting requires restoring from a backup.

##### Gateway Compatibility

- Use gateways version 5.3.0 or newer with the v5.3.0 Control Plane. Older gateways may not function correctly with the updated Tyk OAS API definitions. 

#### Essential steps prior to upgrade

- Back-Up Your Environment;
- Export Tyk OAS API Definitions;
- Consult Upgrade Instructions: For a detailed upgrade process and further steps, refer to the [upgrade instructions guide](#upgrading-tyk).
```

Reviewer 2: @andyo-tyk what do you think?

Author: @lghiur TBC re automatic conversion

Reviewer 3: > - Automatic Conversion: Upgrading automatically updates your Tyk OAS API definitions to the [latest format]({{< ref "tyk-apis/tyk-gateway-api/oas/x-tyk-oas-doc" >}}) used in v5.3. <TBC - Laurentiu??>

- For OSS scenarion, when OAS API definitions are part of the file system, the upgrade doesn't autoamtically happen and users need to update their OAS API Definitions

> - Non-Reversible Change: The format change is one-way within the product. Reverting to your previous version requires restoring from a backup.

- This is only valid, if users upgrade to 5.3.0, they **Create new APIs** on 5.3.0 and then they decide to downgrade

Author: @lghiur Thanks for reviewing. I have updated the suggestion. Does it look ok?

@letzya Are we still moving this to Dashboard Release Notes:

_##### Gateway Compatibility <Should be in Dashboard release notes TODO>
- Use gateways version 5.3.0 or newer with the v5.3.0 Control Plane. Older gateways may not function correctly with the updated Tyk OAS API definitions._

Reviewer 3: @dcs3spp
> OAS API definitions that are part of the file system are not  ...

Maybe just say that for OSS users that store API Definitions on the file system

Tyk OAS APIs using the 5.3.0 API definition may not work with pre-5.3.0 versions of Tyk Gateway and doing so could provoke unpredictable behaviour including potential crashes. We recommend retaining copies of existing Tyk OAS API definitions before upgrading in case you choose to downgrade later.

Users are strongly advised to follow the recommended [upgrade instructions](#upgrading-tyk) provided by Tyk before applying any updates.

##### Python plugin support

Starting from Tyk Gateway version v5.3.0, Python is no longer bundled with the official Tyk Gateway Docker image to reduce exposure to security vulnerabilities in the Python libraries.

> Whilst the Gateway still supports Python plugins, you must [extend the image]({{< ref "plugins/supported-languages/rich-plugins/python/python#install-the-python-development-packages" >}}) to add the language support.

Reviewer 2: ```suggestion
Whilst the Gateway still supports Python plugins, you will need to extend the Dockerfile and [rebuild the image]({{< ref "plugins/supported-languages/rich-plugins/python/python#install-the-python-development-packages" >}}) to use Python plugin with with Tyk Gateway.
```

</details>
</li>
<li>
<details>
<summary>Enhanced OpenTelemetry trace generation for UDG with mixed data sources</summary>

This release introduces an enhanced trace generation system for Universal Data Graph (UDG). It consolidates all spans from both Tyk-managed and external data source executions into a single trace when used together. Furthermore, when UDG solely utilises Tyk-managed data sources, trace management is simplified and operational visibility is improved.
</details>
</li>
<li>
<details>
<summary>Disabled normalise and validate in GraphQL Engine</summary>

For GraphQL requests normalisation and validation has been disabled in the GraphQL engine. Both of those actions were performed in the Tyk Gateway and were unnecessary to be done again in the engine. This enhances performance slightly and makes detailed OTel traces concise and easier to read.
</details>
</li>
<li>
<details>
<summary>Added introspection_config to Tyk Classic API definition</summary>

> Added a new set of settings in the API definition called *introspection_config*. This enables users to have more control over introspection on a single GQL API level.

Reviewer 4: Can we remove this part completely? I mean yes, technically we added this setting but it does nothing yet. It is just in the API definition but has no impact on any behaviour in the GW nor in the Dashboard.

Author: @agata-wit I've removed it in GW release notes and if it is in Dashboard release notes I'll remove it there also

</details>
</li>
<li>
<details>

> <summary>Enhanced UDG's handling of arrays of objects in OpenAPI Documents</summary>

Reviewer 4: ```suggestion
<summary>Enhanced OAS-to-UDG converter handling of arrays of objects in OpenAPI Documents</summary>
```

<ul>
<li>
<details>
<summary>Improved OAuth token management in Redis</summary>

In this release, we fixed automated token trimming in Redis, ensuring efficient management of OAuth tokens by implementing a new hourly job within the Gateway and providing a manual trigger endpoint. 
</details>
</li>
<li>
<details>
<summary>Tyk Gateway now validates RFC3339 Date-Time Formats</summary>

We fixed a bug in the Tyk OAS Validate Request middleware where we were not correctly validating date-time format schema, which could lead to invalid date-time values reaching the upstream services.
</details>
</li>
<li>
<details>
<summary>Inaccurate Distributed Rate Limiting (DRL) behavior on Gateway startup</summary>

Fixed an issue when using the Distributed Rate Limiter (DRL) where the Gateway did not apply any rate limit until a DRL notification was received. Now the rate of requests will be limited at 100% of the configured rate limit until the DRL notification is received, after which the limit will be reduced to an even share of the total (i.e. 100% divided by the number of Gateways) per the rate limit algorithm design.
</details>
</li>
<li>
<details>

> <summary>Duplicate fields added by OAS-to-UDG translator</summary>

Reviewer 4: ```suggestion
<summary>Duplicate fields added by OAS-to-UDG converter</summary>
```

Fixed an issue where the OAS-to-UDG converter was sometimes adding the same field to an object type many times. This caused issues with the resulting GQL schema and made it non-compliant with GQL specification.
</details>
</li>
<li>
<details>
<summary>Gateway issue processing queries with GQL Engine</summary>

Fixed an issue where the Gateway attempted to execute a query with GQL engine version 1 (which lacks OTel support), while simultaneously trying to validate the same query with the OpenTelemetry (OTel) supported engine. It caused the API to fail with an error message "Error socket hang up". Right now with OTel enabled, the gateway will enforce GQL engine to default to version 2, so that this problem doesn't occur anymore.
</details>
</li>
<li>
<details>
<summary>Handling arrays of objects in endpoint responses by OAS-to-UDG converter</summary>

The OAS-to-UDG converter now effectively handles array of objects within POST paths. Previously, there were instances where the converter failed to accurately interpret and represent these structures in the generated UDG configuration.
</details>
</li>
<li>
<details>

> <summary>Playground issues in Cloud/K8s deployments</summary>

Reviewer 4: ```suggestion
<summary>GQL Playground issues related to encoding of request response</summary>
```

An issue was identified where the encoding from the GQL upstream cache was causing readability problems in the response body. Specifically, the upstream GQL cache was utilizing brotli compression and not respecting the Accept-Encoding header. Consequently, larger response bodies became increasingly unreadable for the GQL engine due to compression, leading to usability issues for users accessing affected content. The issue has now been fixed by adding the brotli encoder to the GQL engine.
</details>
</li>
<li>
<details>

> <summary>OAS-to-UDG Converter issue with "json" Return Type</summary>

Reviewer 4: ```suggestion
<summary>OAS-to-UDG converter issue with "JSON" return type</summary>
```

> OAS-to-UDG converter was unable to correctly process Tyk OAS API definitions where "json" was used as one of enum values. This issue is now fixed and whenever "json" is used as one of enums in the OpenAPI description, it will get correctly transformed into a custom scalar in GQL schema.

Reviewer 4: ```suggestion
OAS-to-UDG converter was unable to correctly process Tyk OAS API definitions where "JSON" was used as one of enum values. This issue is now fixed and whenever "JSON" is used as one of enums in the OpenAPI description, it will get correctly transformed into a custom scalar in GQL schema.
```

</details>
</li>
<li>
<details>
<summary>Gateway Panic during API Edit with Virtual Endpoint</summary>

Fixed an issue where the Gateway could panic while updating a Tyk OAS API with the Virtual Endpoint middleware configured.
</details>
</li>
<li>
<details>
<summary>Gateway panics during API Reload with JavaScript middleware bundle</summary>

Fixed an issue where reloading a bundle containing JS plugins could cause the Gateway to panic.
</details>
</li>
<li>
<details>

> <summary>GraphQL introspection issue with Allow List</summary>

Reviewer 4: ```suggestion
<summary>GraphQL introspection issue when Allow/Block List enabled</summary>
```
